src/app/AttributePathExpandIterator-Checked.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/AttributePathExpandIterator-Checked.h>
+
+namespace chip {
+namespace app {
+AttributePathExpandIteratorChecked::AttributePathExpandIteratorChecked(InteractionModel::DataModel * dataModel,
+                                                                       SingleLinkedListNode<AttributePathParams> * attributePath) :
+    mDataModelIterator(dataModel, attributePath),
+    mEmberIterator(dataModel, attributePath)
+{
+    CheckOutputsIdentical("Constructor");
+}
+
+bool AttributePathExpandIteratorChecked::Next()
+{
+    bool dmResult = mDataModelIterator.Next();
+    bool emResult = mEmberIterator.Next();
+
+    CheckOutputsIdentical("Next");
+
+    VerifyOrDie(dmResult == emResult);
+
+    return emResult;
+}
+
+bool AttributePathExpandIteratorChecked::Get(ConcreteAttributePath & aPath)
+{
+    CheckOutputsIdentical("Get");
+    return mEmberIterator.Get(aPath);
+}
+
+void AttributePathExpandIteratorChecked::ResetCurrentCluster()
+{
+    mDataModelIterator.ResetCurrentCluster();
+    mEmberIterator.ResetCurrentCluster();
+
+    CheckOutputsIdentical("ResetCurrentCluster");
+}
+
+void AttributePathExpandIteratorChecked::ResetTo(SingleLinkedListNode<AttributePathParams> * paths)
+
+{
+    mDataModelIterator.ResetTo(paths);
+    mEmberIterator.ResetTo(paths);
+    CheckOutputsIdentical("ResetTo");
+}
+
+void AttributePathExpandIteratorChecked::CheckOutputsIdentical(const char * msg)
+{
+    ConcreteAttributePath dmPath;
+    ConcreteAttributePath emPath;
+
+    bool dmResult = mDataModelIterator.Get(dmPath);
+    bool emResult = mEmberIterator.Get(emPath);
+
+    // NOTE: extra logic because mExpanded is NOT considered in operator== (ugly...)
+    if ((dmResult == emResult) && (dmPath == emPath) && (dmPath.mExpanded == emPath.mExpanded))
```

Comment:
If both results are false, aren't the paths possibly garbage, so they could test different and whatnot?

Reply:
I was unsure here - the code does reset the paths to a known value, it looks like this in AttributePathExpandIterator:

```
    // Reset to default, invalid value.
    mOutputPath = ConcreteReadAttributePath();
    return false;
```

And the updated code matches that. So you are correct that technically if the values are false, the path content should not matter. However at the same time I am making a stronger statement here: the values match even if returning false.

I can change it if you think it is important, however for now I think this has a stronger coverage.

Reply:
This is just making assumptions not guaranteed by the API.  So it could start randomly failing....  I'm OK with that if you are.

Reply:
I expect we will delete the `-Check` interface at some point since I assume we will want to switch to the `data model interface` fully at some point in time. Because of that, I prefer the existing strict bug-by-bug compatibility.

Reply:
Updated to not check patch equivalence on false return values. This seems easier to review and if I had to change code anyway (either comment or this, may as well update it).